### PR TITLE
feat(user-group): bulk group member manager

### PR DESCRIPTION
**Pull request by mediremi**
Closes [DHIS2-10524](https://jira.dhis2.org/browse/DHIS2-10524), [DHIS2-11829](https://jira.dhis2.org/browse/DHIS2-11829) and [DHIS2-11827](https://jira.dhis2.org/browse/DHIS2-11827)

## Changes

- Adds a UI component for bulk managing group members when editing a group, with changes persisted using the collections API
- Replaces user, group and role `PUT` update requests with JSON patch.

## Screenshot

image

### CODE REVIEW

File: src/components/GroupForm/BulkUserManager/BulkUserManager.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        groupId,
+        mode,
+    })
+    const pendingChanges = usePendingChanges()
+
+    useEffect(() => {
+        onChange(
+            pendingChanges.map(({ action, userId }) => ({ action, userId }))
+        )
+    }, [pendingChanges])
```
Comment (mediremi) [inline src/components/GroupForm/BulkUserManager/BulkUserManager.js:35]: `onChange` was not added as a `useEffect` dependency here as React Final Form provides fields with a new `onChange` when their value is updated, which would trigger an infinite render loop in this case.

File: src/components/GroupForm/BulkUserManager/hooks/useMap.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { useState, useMemo } from 'react'
+
+export const useMap = () => {
+    const [map, setMap] = useState(new Map())
+
+    return useMemo(
```
Comment (mediremi) [inline src/components/GroupForm/BulkUserManager/hooks/useMap.js:6]: `useMemo` ensures a that the reference to the 'map' returned by `useMap` is stable unless mutated, allowing the following pattern:

```js
const someMap = useMap()

useEffect(() => {
  // someMap has been mutated, so trigger e.g. onChange handlers
}, [someMap])
```

File: src/components/GroupForm/BulkUserManager/hooks/useUsers/index.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import { useSet } from './useSet'
+
+const FILTER_DEBOUNCE_MS = 375
+
+export const useUsers = ({ groupId, mode }) => {
+    const queries = useQueries({ groupId })
+    const membersDataQuery = useDataQuery(queries.members, { lazy: true })
+    const nonMembersDataQuery = useDataQuery(queries.nonMembers, { lazy: true })
+    const [prevMembers, setPrevMembers] = useState()
+    const [prevNonMembers, setPrevNonMembers] = useState()
```
Comment (mediremi) [inline src/components/GroupForm/BulkUserManager/hooks/useUsers/index.js:14]: Keep track of previous values of `members` and `nonMembers` in order to render them whilst the next/previous page of users is loading. Once https://github.com/dhis2/app-runtime/pull/1050 is merged, this logic can be eliminated.

**Comment by mediremi** [comment]
Converting to draft as work needs to be done refactoring the bulk user manager so that it is ready to merge into `@dhis2/ui`.

File: src/components/BulkMemberManager/BulkMemberManager.js [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    addEntity,
+    removeEntity,
+    cancelAddEntity,
+    cancelRemoveEntity,
+} from './pendingChangesActions'
+import PendingChangesPropType from './PendingChangesPropType'
+import ResultsTable from './ResultsTable'
+import TopBar from './TopBar'
+
+const BulkMemberManager = ({
```
Comment (mediremi) [inline src/components/BulkMemberManager/BulkMemberManager.js:19]: This component will be extracted from this app and merged into `@dhis2/ui` in future.

File: src/components/BulkMemberManager/BulkMemberManager.js [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    className,
+    canManageMembers,
```
Comment (mediremi) [inline src/components/BulkMemberManager/BulkMemberManager.js:21]: Not too happy with the name of this prop but not sure what would work better :thinking: `canManageExistingMembers`?

File: src/components/GroupForm/BulkUserManager/BulkUserManager.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+              selectedCount,
+          })
+
+const renderRowActionLabel = ({ mode }) =>
+    mode === 'MEMBERS' ? i18n.t('Remove from group') : i18n.t('Add to group')
+
+const transformUserResults = results =>
+    results.users.map(user => ({
+        ...user,
+        displayName: user.username || user.name,
```
Comment (mediremi) [inline src/components/GroupForm/BulkUserManager/BulkUserManager.js:39]: This fallback is only needed in development as our default database used by the play/debug instances contains some users without usernames

File: src/components/GroupForm/BulkUserManager/BulkUserManager.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        firstItem,
+        lastItem,
+        total,
+    })
+
+const BulkUserManager = ({ className, groupId, value, onChange }) => {
+    const { membersQuery, nonMembersQuery } = useQueries({ groupId })
+
+    return (
+        <BulkMemberManager
```
Comment (mediremi) [inline src/components/GroupForm/BulkUserManager/BulkUserManager.js:58]: This isn't the easiest component to use boilerplate-wise so we should ensure we provide good and well commented examples in our UI documentation (once merged into `@dhis2/ui`).

File: src/components/BulkMemberManager/hooks/useResults/useResults.js [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    const [filter, setFilter] =
+        mode === 'MEMBERS'
+            ? [membersFilter, setMembersFilter]
+            : [nonMembersFilter, setNonMembersFilter]
+
+    return {
+        loading: !called || loading,
+        error,
+        results,
+        pager: {
```
Comment (mediremi) [inline src/components/BulkMemberManager/hooks/useResults/useResults.js:84]: `BulkMemberManager` assumes that the queries provided return paginated data. I think this is a fair assumption but maybe we should have an assertion somewhere that this is the case? That could allow us to provide a better error message if the consumer of this component does not realise that they passed unpaginated queries?

**Review by HendrikThePendric — APPROVED** [review]
Awesome work. It's great that users can now manage users in a user-group again. The component is also abstracted very neatly. I think it's pretty much ready to go into @dhis2/ui as-is. But we will defer that until another app needs this type of UI.

Thanks for this PR, als all the other great work you've fone for DHIS2 @mediremi.

**Before merging this in, could you please address the two very minor comments I have left in the code?**

Also, I think you can remove the warning about not being able to add and remove at the same time. @vietnguyen has merged a fix for that into master earlier on today.

**Pull request merged**